File: app/src/main/java/com/example/getfood/ui/base/BasePresenter.java
package com.example.getfood.ui.base;

import com.example.getfood.api.FireBaseApiManager;
import com.example.getfood.data.DataManager;

public abstract class BasePresenter<V extends BaseView> implements BaseMvpPresenter<V> {

    private static final String TAG = "BasePresenter";
    protected FireBaseApiManager apiManager;
    protected DataManager dataManager;
    private V mMvpView;

    public BasePresenter() {
    }

    @Override
    public void onAttach(V mvpView) {
        mMvpView = mvpView;
        apiManager = FireBaseApiManager.getInstance();
        dataManager = DataManager.getInstance();
    }

    public FireBaseApiManager getApiManager() {
        return apiManager;
    }

    @Override
    public void onDetach() {
// mCompositeDisposable.dispose();
        mMvpView.hideLoading();
        mMvpView = null;
    }

    public boolean isViewAttached() {
        return mMvpView != null;
    }

    public V getMvpView() {
        return mMvpView;
    }

    public void checkViewAttached() {
        if (!isViewAttached()) throw new MvpViewNotAttachedException();
    }

    public static class MvpViewNotAttachedException extends RuntimeException {
        public MvpViewNotAttachedException() {
            super("Please call Presenter.onAttach(MvpView) before" +
                    " requesting data to the Presenter");
        }
    }
}

File: app/src/main/java/com/example/getfood/ui/base/BasePresenter.kt
package com.example.getfood.ui.base

import com.example.firebase_api_library.api.FireBaseApiManager
import com.example.firebase_api_library.api.FireBaseApiManager.Companion.instance
import com.example.getfood.data.DataManager

abstract class BasePresenter<V : BaseView?> : BaseMvpPresenter<V> {
    var apiManager: FireBaseApiManager? = null
        protected set
    protected lateinit var dataManager: DataManager
    var mvpView: V? = null
        private set

    override fun onAttach(mvpView: V) {
        this.mvpView = mvpView
        apiManager = instance
        dataManager = DataManager.instance
    }

    override fun onDetach() {
// mCompositeDisposable.dispose();
        mvpView!!.hideLoading()
        mvpView = null
    }

    val isViewAttached: Boolean
        get() = mvpView != null

    fun checkViewAttached() {
        if (!isViewAttached) throw MvpViewNotAttachedException()
    }

    class MvpViewNotAttachedException : RuntimeException("Please call Presenter.onAttach(MvpView) before" +
            " requesting data to the Presenter")

    companion object {
        private const val TAG = "BasePresenter"
    }
}